pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+	disruptionutils "sigs.k8s.io/karpenter/pkg/utils/disruption"
+)
+
+const (
+	maxConcurrentNominations = 5
+	maxBatchSize             = 100
+
+	singleTimeout = 3 * time.Minute
+	multiTimeout  = 1 * time.Minute
+)
+
+type Controller struct {
+	kubeClient             client.Client
+	cluster                *state.Cluster
+	recorder               events.Recorder
+	provisioner            *provisioning.Provisioner
+	cloudProvider          cloudprovider.CloudProvider
+	clk                    clock.Clock
+	instanceTypes          map[string]map[string]*cloudprovider.InstanceType
+	lastConsolidationState time.Time
+}
+
+func NewController(clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster,
+) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		recorder:      recorder,
+		provisioner:   provisioner,
+		clk:           clk,
+		cloudProvider: cp,
+	}
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+func (c *Controller) Name() string {
+	return "nodeclaim.underutilization"
+}
+
+// IsConsolidated returns true if nothing has changed since markConsolidated was called.
+func (c *Controller) IsConsolidated() bool {
+	return c.lastConsolidationState.Equal(c.cluster.ConsolidationState())
+}
+
+// markConsolidated records the current state of the cluster.
+func (c *Controller) markConsolidated() {
+	c.lastConsolidationState = c.cluster.ConsolidationState()
+}
+
+// Reconcile will try to find consolidation decisions in the cluster:
+//  1. First get the active consolidation decisions in the cluster and see if any decisions
+//     can be invalidated, removing the Underutilization condition from the nodeclaims if invalid.
+//  2. If there are no actively progressing consolidations, do not do anything until cluster
+//     state is no longer in a consolidated state. This is important to reduce resource usage
+//     when the cluster is idle.
+//  3. If there are 5 or more actively progressing consolidations, exit early, as we should just
+//     wait for these to finish.
+//  4. Compute a consolidation decision with scheduling simulations
+//  5. If we find a valid decision, mark all node claims as Underutilized.
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	noms := c.getActiveConsolidations(ctx)
+	// If there are no active nominations and it's been less than 5 minutes since the last time
+	// the cluster was marked as consolidated, wait.
+	if len(noms) == 0 && c.IsConsolidated() {
+		return reconcile.Result{RequeueAfter: time.Until(c.cluster.ConsolidationState().Add(5 * time.Minute))}, nil
+	}
+	// If there are too many nominations, see if we can quickly invalidate any existing nominations
+	if len(noms) >= maxConcurrentNominations {
```

Comment:
```suggestion
	if len(noms) > maxConcurrentNominations {
```

Just greater? I'd assume that max is inclusive